0-module-and-module-source.md

```diff
@@ -333,28 +410,80 @@ This proposal does not impose on host-defined import behavior.
+  when it is present.
+- The origin of the source, the basis for a host deciding whether to allow the
+  module to be evaluated when a Content-Security-Policy is in place.
+  The origin will often be identical to the origin of `import.meta.url`.
+  The origin will exist in host data of the module source and must
+  not be virtualizable, lest a virtual host be able to escape a same origin
+  policy.
+
+So, although these values are often related, they must be independent features.
+Virtual host implementations cannot necessarily rely on the `import.meta.url`
+to exist and virtual host implementations must be able to virtualize the
+referrer.
+
+The authors of this proposal vacilated between versions that made the referrer
+explicit or implicit.
+All of the authors shared a goal to make the `Module` and `ModuleSource`
+features incrementally explainable.
+Since it is possible to construct a trivial module system without the concept
+of a referrer, a graduated tutorial introducing this feature (like this
+explainer!) should be able to defer or formally ignore referral until it
+becomes necessary to explain relative module specifier resolution.
+
+To that end, we agreed that the referrer should be optional and appear later
+in function signatures than other more essential parameters, if at all.
+We relegated the `referrer` to at least be a property of the `Module`
+constructor's options bag.
+We then realized that the options bag was more analogous to the `Proxy` handler
+object, such that the virtual host could choose to extend the handler with any
+property of any type to carry the referrer, such that an import hook might
+resolve a relative specifier using `this.referrer` (a `string` with the
+module's own specifier), `this.slot` (a `number` representing the index into an
+array of pre-computed resolutions for a bundle), or whatever other protocol the
+virtual host chose to implement.
+
+This was only possible because host-defined resolution behavior and
+virtual-host-defined resolution behavior do not need to communicate.
+Host-defined modules resolve import specifiers based on information in the
+referring module record's host data internal slot.
+At no point does a host-defined module need to consult the referrer of a
+virtual-host-defined module.
+
+We additionally elected not to repeat what we believe to be a mistake in the
+design of `Proxy`.
+The `Module` constructor eagerly captures all the optional methods of the
+module handler rather than accessing them afresh for each invocation.
+So, we sacrifice some dynamism but still thread the handler as the target
+object for all method invocations.
+
+A consequence of this design is that any `Module` instance will retain a
+`handler`, where an options bag would be immediately released to the GC
+nursery.
+And, if the handler carries a `referrer`, every `Module` instance needs a
+unique handler instance, even though many module instances can share
+the same hooks.
```

Comment:
Oh might be interesting to actually write one of the examples in this form instead of the "class" approach.